### PR TITLE
Migrate SmoothQuant for IPEX to 3.x API

### PR DESCRIPTION
**Pull request by violetch24**
## Type of Change

Migrate SmoothQuant for IPEX to 3.x API, API changed

## Description
To-do:
- [x] Enable different scales for Linears sharing the same input
- [x] Add more test cases (autotune)
- [x] Add accuracy test

## How has this PR been tested?
UT passed

## Dependency Change?

### CODE REVIEW

File: neural_compressor/torch/quantization/algorithm_entry.py [hunk 1]
```diff
@@ -156,6 +157,62 @@ def static_quant_entry(
...
+###################### Smooth Quant Algo Entry ##################################
+@register_algo(name=SMOOTH_QUANT)
+@torch.no_grad()
+def smooth_quant_entry(
+    model: torch.nn.Module, configs_mapping: Dict[Tuple[str, callable], SmoothQuantConfig], *args, **kwargs
+) -> torch.nn.Module:
+    logger.info("Quantize model with the smooth quant algorithm.")
+    from neural_compressor.torch.algorithms.smooth_quant.smooth_quant import smooth_quantize
+
+    # rebuild tune_cfg for smooth_quant function
```
Comment (yiliu30) [inline neural_compressor/torch/quantization/algorithm_entry.py:169]:
```suggestion
    # Convert the user config into internal format.
```
nit: Quantization should not aware tuning-related matters. This comment may confuse the user.

File: test/3x/torch/quantization/test_smooth_quant.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import copy
+
+import intel_extension_for_pytorch as ipex
```
Comment (yiliu30) [inline test/3x/torch/quantization/test_smooth_quant.py:3]: Suggest importing it when `is_ipex_available()` is `True`.

File: neural_compressor/torch/algorithms/smooth_quant/utility.py [hunk 3]
```diff
@@ -0,0 +1,2591 @@
...
+        [op_type, sum(res[op_type].values()), res[op_type]["INT8"], res[op_type]["BF16"], res[op_type]["FP32"]]
+        for op_type in res.keys()
+    ]
+
+    Statistics(
+        output_data, header="Mixed Precision Statistics", field_names=["Op Type", "Total", "INT8", "BF16", "FP32"]
+    ).print_stat()
+
+
+def get_quantizable_ops_recursively(model, example_inputs):
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/smooth_quant/utility.py:561]: I noticed the same function [here](https://github.com/intel/neural-compressor/blob/2a86aeafc754ca3b7495138381efb8faa9397fdf/neural_compressor/torch/algorithms/static_quant/utility.py#L375). If they indeed serve the same purpose, please reuse the code. Also, make sure to review other functions.

**Review by yiliu30 — APPROVED** [review]

**Review by yiliu30 — APPROVED** [review]
Left some comments, others are LGTM.

File: neural_compressor/torch/algorithms/smooth_quant/utility.py (hunk 3 above)
Comment (xin3he) [reply neural_compressor/torch/algorithms/smooth_quant/utility.py]: @violetch24  I saw you moved them to `torch/utils/utility.py`, which is not so reasonable. this utility aims to provide userfacing utilities while these functions are pure for intenal usage and only for IPEX. 
@yiliu30 I prefer to keep them in static_quant folder and import them to smooth_quant folder. Per my understanding, smooth quant is an enhancement of static quant, so we make static quant as the base.

**Comment by xin3he** [comment]
@violetch24  Unittest is not enough, please migrate uts pure for algorithms and more sq accuracy check.

File: test/3x/torch/quantization/test_smooth_quant.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        ],
+    )
+    def test_sq_linear_params(self, act_sym, act_algo, alpha, folding, scale_sharing):
+        fp32_model = copy.deepcopy(model)
+        quant_config = SmoothQuantConfig(
+            act_sym=act_sym, act_algo=act_algo, alpha=alpha, folding=folding, scale_sharing=scale_sharing
+        )
+        example_inputs = torch.randn([1, 3])
+        q_model = quantize(fp32_model, quant_config=quant_config, run_fn=run_fn, example_inputs=example_inputs)
+        assert q_model is not None, "Quantization failed!"
```
Comment (xin3he) [inline test/3x/torch/quantization/test_smooth_quant.py:62]: It's better to have accuracy check here.

File: test/3x/torch/quantization/test_smooth_quant.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    @pytest.mark.skipif(not is_ipex_available(), reason="Requires IPEX")
+    def test_sq_accuracy(self):
+        fp32_model = copy.deepcopy(model)
+        quant_config = SmoothQuantConfig(act_sym=True, act_algo="kl", alpha="auto", folding=True, scale_sharing=True)
+        example_inputs = torch.zeros([1, 3])
+        q_model = quantize(fp32_model, quant_config=quant_config, run_fn=run_fn, example_inputs=example_inputs)
+        assert q_model is not None, "Quantization failed!"
+        output1 = model(example_inputs)
+        output2 = q_model(example_inputs)
+        assert torch.allclose(output1, output2, atol=2e-2), "Accuracy gap atol > 0.02 is unexpected. Please check."
```
Comment (xin3he) [inline test/3x/torch/quantization/test_smooth_quant.py:88]: Please add them back: https://github.com/intel/neural-compressor/blame/master/test/algorithm/test_smooth_quant.py
Tests without INC quantization should be put in `test/3x/torch/algorithms/test_smoothquant.py`

**Review by xin3he — CHANGES_REQUESTED** [review]

**Comment by xin3he** [comment]
Currently, we have enabled save & load API. Please refer to [fp8 UT's test_save_load](https://github.com/intel/neural-compressor/blob/3fa9ab1fc5d0c9cdfad9b158e9a23f96bb80ce4f/test/3x/torch/quantization/habana_fp8/test_fp8.py#L152)

**Review by xin3he — APPROVED** [review]
There will be another PR to polish SmoothQuant UTs.
